Add Enhanced Dev Preview documentation (#8)
openshift/network-edge-tools: docs/blogs/EnhancedDevPreviewGatewayAPI/GettingCreative.md

# Getting Creative

In the previous section, we showed how to create a simple HTTPRoute. Let’s expand upon this example and include more
advanced usages of Gateway API.

## Route Matching

> Gateway API has a powerful API for directing traffic based on request characteristics such as path, headers, query

Reviewer: ```suggestion
Gateway API is a powerful API for directing traffic based on request characteristics such as path, headers, query
```
or
```suggestion
Gateway API provides a powerful API for directing traffic based on request characteristics such as path, headers, query
```

Author: Done.

parameters, and HTTP methods. Let’s explore how we can use these functions to direct traffic to different backends.

#### Path Based Routing
In this first example, we will match traffic based on the paths `/foo` and `/bar`. Path-based routing is useful for
segmenting a website into various backend services. In the current OpenShift Route API, we must make a Route per path:
```yaml
apiVersion: v1
kind: Route
metadata:
  name: http-foo
  namespace: demo-app
spec:

>   host: http.gwapi.${DOMAIN}

Reviewer: This host should match the other route, correct?  And the OpenShift Route object shouldn't be using a domain with "gwapi" in it.  
```suggestion
  host: http.${DOMAIN}
```

Author: Whoops yea there is a mistake here. I was describing the conversion that will happen quite literally: i.e. the Route and the Gateway/HTTPRoute use the same hostnames, but I can remove the "gwapi", I think folks will get the point either way.

  to:
    kind: Service
    name: foo-app
  port:
    targetPort: 8080
  path: /foo
---
apiVersion: v1
kind: Route
metadata:
  name: http-bar
  namespace: demo-app
spec:
  host: http.${DOMAIN}
  to:
    kind: Service
    name: bar-app
  port:
    targetPort: 8080
  path: /bar
```

However, Gateway API enables the path-based matching to be declared in a single HTTPRoute object as we will demonstrate
in this example.

> First, let’s create another demo application deployment and service using `oc new-app`, our `cakephp-ingress-demo`
> project, and the `bar` git branch. This will be used as an alternate server backend, `bar-app`, to direct traffic to:
> ```shell
> $ oc new-app -n demo-app --name bar-app https://github.com/openshiftdemos/cakephp-ingress-demo#bar
> $ oc rollout status deployment -w -n demo-app bar-app
> ```

Reviewer: ````suggestion
First, let’s create another demo application deployment and service using the `oc new-app` command and the `bar` branch of our `cakephp-ingress-demo` GitHub
project. This will be used as an alternate server backend, `bar-app`, to direct traffic to:
```console
$ oc new-app -n demo-app --name bar-app https://github.com/openshiftdemos/cakephp-ingress-demo#bar
$ oc rollout status deployment -w -n demo-app bar-app
[expected output]
```
````
What is the expected output?

Author: Added expected output (to both, just to be consistent).

Next, let’s modify our existing HTTPRoute to direct traffic based on path. We will add a `/foo` path match on the

> existing rule for the `foo-app` application as well as a new rule with a `/bar` path match and a backendRef pointing to the

Reviewer: Maybe it would be better to spell out "backend reference":
```suggestion
existing rule for the `foo-app` application as well as a new rule with a `/bar` path match and a backend reference pointing to the
```
Alternatively, you could use ``` `backendref` ```.

Author: Done.

`bar-app` application:
```yaml
$ DOMAIN=$(oc get ingresses.config/cluster -o jsonpath={.spec.domain})
$ oc apply -f - <<EOF
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: http
  namespace: demo-app
spec:
  parentRefs:
  - name: demo-gateway
    namespace: openshift-ingress
  hostnames: ["http.gwapi.${DOMAIN}"]
  rules:
  - matches: #[1]
    - path:
        type: Exact
        value: /foo
    backendRefs:
    - name: foo-app
      port: 8080
  - matches: #[2]
    - path:
        type: Exact
        value: /bar
    backendRefs:
    - name: bar-app
      port: 8080
EOF
```
[1] `/foo` path match on the existing rule for the `foo-app` application.\

> [2] new rule with a `/bar` path match and a backendRef pointing to the `bar-app` application.

Reviewer: ```suggestion
[2] New rule with a `/bar` path match and a `backendRef` pointing to the `bar-app` application.
```

Author: Done.

Now let’s test our path-based routing configuration:

> ```shell

Reviewer: ````suggestion
```console
````

Author: Done.

$ curl -s http://http.gwapi.${DOMAIN}/foo
Welcome to the foo application!

$ curl -s http://http.gwapi.${DOMAIN}/bar
Welcome to the bar application!
```

### Other Matching Functions
Gateway API offers additional ways of routing traffic. Besides matching on path, Gateway API also allows you to match

> requests based on HTTP headers, URL query parameters, and HTTP method. Currently, the OpenShift Route API does not offer
> a way to match traffic on any of these options.

Reviewer: ```suggestion
requests based on HTTP headers, URL query parameters, and HTTP methods. Currently, the OpenShift Route API does not offer
a way to match traffic on any of these attributes.
```

Author: Done.

The following example shows header and URL query parameter matching. First, we set a match rule using `queryParams` that
if the `app` query parameter is set to `foo`, the request should be forwarded to the `foo-app` application; and second, we
set a match rule using headers that if the `x-app` header is set to `bar`, the request should be forwarded to the `bar-app`
application:
```yaml
$ oc apply -f - <<EOF
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: http
  namespace: demo-app
spec:
  parentRefs:
  - name: demo-gateway
    namespace: openshift-ingress
  hostnames: ["http.gwapi.${DOMAIN}"]
  rules:
  - matches:
    - queryParams: #[1]
      - name: app
        value: foo
    backendRefs:
    - name: foo-app
      port: 8080
  - matches:
    - headers: #[2]
      - name: x-app
        value: bar 
    backendRefs:
    - name: bar-app
      port: 8080
EOF
```
[1] `queryParams`: if the `app` query parameter is set to `foo`, the request should be forwarded to the `foo-app` application.\
[2] `headers`: if the `x-app` header is set to `bar`, the request should be forwarded to the `bar-app` application.

Now let’s test query parameter matching:

> ```shell

Reviewer: ````suggestion
```console
````

Author: Done.

$ curl -s http://http.gwapi.${DOMAIN}?app=foo
Welcome to the foo application!
```

Next, let's test header matching:

> ```shell

Reviewer: ````suggestion
```console
````

Author: Done.

$ curl -sH "x-app: bar" http://http.gwapi.${DOMAIN}
Welcome to the bar application!
```

You’ve successfully configured Gateway API to match traffic based on path, query parameters, and headers.

## TLS Termination with Gateway API

> In OpenShift, we have the concept of [termination types](https://docs.openshift.com/container-platform/4.11/networking/routes/secured-routes.html) 

Reviewer 2: Maybe update to a 4.13 version of the docs.

Author: Good idea. Done.

> In OpenShift, we have the concept of [termination types](https://docs.openshift.com/container-platform/4.13/networking/routes/secured-routes.html) 

Reviewer: ```suggestion
In OpenShift, we have the concept of [TLS termination types](https://docs.openshift.com/container-platform/4.13/networking/routes/secured-routes.html) 
```

Author: Done.

> to achieve different forms of secured routing. In this example, we will explore how to create an edge terminated route with Gateway API.

Reviewer: ```suggestion
to achieve different forms of secured routing. In this example, we will explore how to create an edge-terminated TLS route with Gateway API.
```

Author: Done.

An edge route in OpenShift terminates TLS encryption at the Ingress Controller before forwarding traffic to the
destination pod:
```yaml
apiVersion: v1
kind: Route
metadata:
  name: http-foo
  namespace: demo-app
spec:

>   host: edge.gwapi.${DOMAIN}

Reviewer: Why would the OpenShift Route object have a host name with "gwapi" in it?
```suggestion
  host: edge.${DOMAIN}
```

Author: I was taking the conversion literally as mentioned above. Updated.

  to:
    kind: Service
    name: foo-app
  port:
    targetPort: 8080
tls:
  termination: edge
  insecureEdgeTerminationPolicy: Redirect
  certificate: <CERT>
  key: <KEY>
path: /foo
```

> In Gateway API, we can translate this concept to terminating the TLS encryption at Gateway before forwarding traffic to

Reviewer: ```suggestion
In Gateway API, we can translate this concept to terminating the TLS encryption at the Gateway before forwarding traffic to
```
It irks me slightly that we're conflating the Gateway (an API object) and the proxy, but I guess it's clear enough what you mean.

Author: I see. The language here is pretty vague, but I'm pretty sure the Gateway API docs uses a pretty similar style in talking about the Gateway. https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io%2fv1beta1.GatewayTLSConfig

Working with the session persistence GEP, I struggle to conceptualize what a Gateway and Route (i.e. HTTPRoute) really is. Sometimes the ideas confuse me when translated back into real configuration, such as Envoy configuration.

Anyways, I added some clarification that it's getting forwarded to an upstream server via an HTTPRoute configuration.

the HTTPRoute.

First, we will need a certificate for configuring the TLS termination. Acquire a certificate with the common name that
you want to use from your CA.  

For testing purposes, you can create your own CA and serving certificate as follows:
```
$ mkdir certs
$ openssl req -x509 -sha256 -nodes -days 365 -newkey rsa:2048 -subj "/O=MyCompany/CN=gwapi.${DOMAIN}" -keyout certs/ca.key -out certs/ca.crt
$ openssl req -out certs/edge.csr -newkey rsa:2048 -nodes -keyout certs/edge.key -subj "/CN=*.gwapi.${DOMAIN}/O=MyCompany"
$ openssl x509 -req -sha256 -days 365 -CA certs/ca.crt -CAkey certs/ca.key -set_serial 0 -in certs/edge.csr -out certs/edge.crt
```

Add the certificate as a secret in the `openshift-ingress` namespace:

> ```shell
> $ oc create -n openshift-ingress secret tls edge-cert --key=certs/edge.key --cert=certs/edge.crt
> ```

Reviewer: ````suggestion
```console
$ oc create -n openshift-ingress secret tls edge-cert --key=certs/edge.key --cert=certs/edge.crt
route.route.openshift.io/edge-cert created
```
````

Author: Done.

Certificates are handled differently in Gateway API compared to OpenShift Route objects. Gateway API uses references to

> Kubernetes Secret objects whereas the OpenShift Route API embeds the certificate and key strings in the Route object.

Reviewer: Please capitalize "secret" or "Secret" consistently.

Author: Done.

> Next, let’s configure our Gateway to terminate TLS. We will add an additional Listener with the same wildcard domain

Reviewer: Why is "Listener" capitalized?

Author: I was trying to refer to a proper API field, a proper noun, but based on https://gateway-api.sigs.k8s.io/api-types/gateway/, they don't capitalize, so I won't.

using port 443 to our existing Gateway from the example above:

> ```yaml

Reviewer: ````suggestion
```console
````

Author: Done.

$ oc apply -f - <<EOF
apiVersion: gateway.networking.k8s.io/v1beta1
kind: Gateway
metadata:
  name: demo-gateway
  namespace: openshift-ingress
spec:
  gatewayClassName: openshift-default
  listeners:
  - name: http
    hostname: "*.gwapi.${DOMAIN}"
    port: 80
    protocol: HTTP
    allowedRoutes:
      namespaces:
        from: All
  - name: edge
    hostname: "*.gwapi.${DOMAIN}"
    port: 443
    protocol: HTTPS
    tls:
      mode: Terminate
      certificateRefs:
      - name: edge-cert
    allowedRoutes:
      namespaces:
        from: All
EOF
```
Next, let’s create the HTTPRoute that will route our TLS terminated connection to the backend:
```yaml
$ oc apply -f - <<EOF
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: edge
  namespace: demo-app
spec:
  parentRefs:
  - name: demo-gateway
    namespace: openshift-ingress
  hostnames: ["edge.gwapi.${DOMAIN}"]
  rules:
  - backendRefs:
    - name: foo-app
      port: 8080
EOF
```
Now test the route we just created over port 443:

> ```shell

Reviewer: ````suggestion
```console
````

Author: Done.

Welcome to the foo application!
```

Next, let's test header matching:
```console
$ curl -sH "x-app: bar" http://http.gwapi.${DOMAIN}
Welcome to the bar application!
```

You’ve successfully configured Gateway API to match traffic based on path, query parameters, and headers.

## TLS Termination with Gateway API
In OpenShift, we have the concept of [TLS termination types](https://docs.openshift.com/container-platform/4.13/networking/routes/secured-routes.html)
to achieve different forms of secured routing. In this example, we will explore how to create an edge-terminated TLS route with Gateway API.

An edge route in OpenShift terminates TLS encryption at the Ingress Controller before forwarding traffic to the
destination pod:
```yaml
apiVersion: v1
kind: Route
metadata:
  name: http-foo
  namespace: demo-app
spec:
  host: edge.${DOMAIN}
  to:
    kind: Service
    name: foo-app
  port:
    targetPort: 8080
tls:
  termination: edge
  insecureEdgeTerminationPolicy: Redirect
  certificate: <CERT>
  key: <KEY>
path: /foo
```
In Gateway API, we can translate this concept to terminating the TLS encryption at the Gateway before forwarding traffic

> to the upstream server via the HTTPRoute configuration.

Reviewer: Better to avoid ambiguous words like "upstream", which different implementations use to mean opposite things.
```suggestion
to the destination pod via the HTTPRoute configuration.
```

Author: Done
